### PR TITLE
Update Program Service to provide all metadata

### PR DESCRIPTION
**Pull request by mikedikan**
@edx/ecommerce for review

JIRA Subtask: https://openedx.atlassian.net/browse/ECOM-5096

### CODE REVIEW

File: course_discovery/apps/api/tests/test_serializers.py [hunk 1]
```diff
@@ -244,8 +278,30 @@ def test_data(self):
...
+            'min_hours_effort_per_week': None,
+            'overview': None,
+            'price_ranges': [],
+            'start': serialize_datetime_to_millis(program.start),
+            'status': program.status,
+            'subjects': [],
+            'transcript_languages': [],
         }
 
+        self.maxDiff = None
```
Comment (clintonb) [inline course_discovery/apps/api/tests/test_serializers.py]: Remove. This is only needed for debugging.

File: course_discovery/apps/api/tests/test_serializers.py [hunk 2]
```diff
@@ -278,15 +334,38 @@ def test_with_exclusions(self):
...
```
Comment (clintonb) [inline course_discovery/apps/api/tests/test_serializers.py]: Remove.

File: course_discovery/apps/api/v1/tests/test_views/test_programs.py [hunk 3]
```diff
@@ -58,11 +68,14 @@ def test_filter_by_uuids(self):
         """ Verify that the endpoint filters programs to those matching the provided UUIDs. """
         url = self.list_path + '?uuids='
 
-        programs = [ProgramFactory(), self.program]
```
Comment (clintonb) [inline course_discovery/apps/api/v1/tests/test_views/test_programs.py:61]: Why is this change necessary given that we are simply testing filtering?

Comment (clintonb) [inline course_discovery/apps/api/v1/tests/test_views/test_programs.py:61]: Actually, why is this file being modified at all?

**Comment by clintonb** [comment]
👍 (once tests pass and comments addressed)

File: course_discovery/apps/api/v1/tests/test_views/test_programs.py (hunk 3 above)
Comment (mikedikan) [reply course_discovery/apps/api/v1/tests/test_views/test_programs.py:61]: The tests in test_programs were failing because the additional metadata fields in the program serializer (specifically, the 'start' property addition) required access to the context to grab the user from the request within the CourseRunSerializer (specifically, the 'get_marketing_url' MethodField) which the 'start' property reads from.

Comment (clintonb) [reply course_discovery/apps/api/v1/tests/test_views/test_programs.py:61]: Got it. The additional context is fine. I don't see a need to add associated courses to these tests.

**Comment by mikedikan** [comment]
@clintonb comments addressed - you were right about courses - I think that was a remnant from an earlier debug theory

File: course_discovery/apps/api/tests/test_serializers.py [hunk 4]
```diff
@@ -39,6 +41,22 @@ def make_request():
     return request
 
 
+def serialize_datetime(d):
+    return d.strftime('%Y-%m-%dT%H:%M:%S') if d else None
+
+
+def serialize_datetime_to_millis(d):
```
Comment (schenedx) [inline course_discovery/apps/api/tests/test_serializers.py]: I am surprised we need the datetime serialized to milli-seconds. I think we should serialize the same datetime format for `program.start` vs `courserun.start`. The cause for the new function is that in the `Programs` serializer, the `start` is serialized using `serializers.DateTimeField()`, while in courseRunSerializer, the `start` is serialized by default.

Comment (clintonb) [reply course_discovery/apps/api/tests/test_serializers.py]: Nice catch. I agree with @schenedx. If you remove the `DateTimeField` for the start field in `ProgramSerializer` you won't need this function.

**Comment by schenedx** [comment]
@mikedikan Just left 1 comment, and rest LGTM.
Another PR suggestion I want to make is, please attach the ticket number to the PR. I think the team like the ticket number in the description of the PR. I personally put ticket number in the branch name, PR title and sometimes, the description of the PR.

**Comment by clintonb** [comment]
Having the ticket in the Git commit's description will ensure the commit is properly linked to the ticket in JIRA as well.

File: course_discovery/apps/api/serializers.py [hunk 5]
```diff
@@ -131,6 +140,24 @@ class Meta(object):
         fields = ('src', 'description', 'image',)
 
 
+class EndorsementSerializer(serializers.ModelSerializer):
+    """Serializer for the ``Endorsement`` model."""
+
+    class Meta(object):
+        model = Endorsement
+        fields = ('endorser', 'quote',)
```
Comment (clintonb) [inline course_discovery/apps/api/serializers.py:157]: endorser should use the `PersonSerializer`.

File: course_discovery/apps/api/v1/tests/test_views/test_programs.py [hunk 6]
```diff
@@ -1,5 +1,6 @@
 from django.core.urlresolvers import reverse
 from rest_framework.test import APITestCase
+from rest_framework.test import APIRequestFactory
```
Comment (clintonb) [inline course_discovery/apps/api/v1/tests/test_views/test_programs.py]: This should be combined with the import above.

File: course_discovery/apps/course_metadata/tests/factories.py [hunk 7]
```diff
@@ -200,6 +244,7 @@ class Meta(object):
     card_image_url = FuzzyText(prefix='https://example.com/program/card')
     partner = factory.SubFactory(PartnerFactory)
     credit_redemption_overview = FuzzyText()
+    category = FuzzyText()
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/tests/factories.py]: Remove. This field is going away. Use the `type` field.

File: course_discovery/apps/api/serializers.py [hunk 8]
```diff
@@ -331,9 +376,14 @@ def get_courses(self, program):
 
     class Meta:
         model = Program
-        fields = ('uuid', 'title', 'subtitle', 'type', 'marketing_slug', 'marketing_url', 'card_image_url',
-                  'banner_image', 'banner_image_url', 'authoring_organizations', 'credit_redemption_overview',
-                  'courses',)
+        fields = (
+            'uuid', 'title', 'subtitle', 'category', 'type', 'status', 'marketing_slug', 'marketing_url', 'courses',
```
Comment (clintonb) [inline course_discovery/apps/api/serializers.py]: Remove the `category` field. It's going away.

File: course_discovery/apps/api/tests/test_serializers.py (hunk 4 above)
Comment (mikedikan) [reply course_discovery/apps/api/tests/test_serializers.py]: When i attempted leaving it using default - it serialized to datetime(x, y, z) which did not match the output - default did not work here - i'm open to other suggestions, but didn't feel like burning up time in a snipe hunt.

Comment (clintonb) [reply course_discovery/apps/api/tests/test_serializers.py]: That's because it's a property rather than a model field. DRF cannot determine the type. This property actually doesn't need to be exposed. It is not publicly displayed anywhere. I don't recall why I added it.

**Comment by mikedikan** [comment]
Latest comments addressed and added JIRA link - please review @clintonb

File: course_discovery/apps/api/serializers.py [hunk 9]
```diff
@@ -317,6 +345,24 @@ class ProgramSerializer(serializers.ModelSerializer):
...
+        many=True, read_only=True, slug_field='code',
+        help_text=_('Languages that course runs in this program are offered in.'),
+    )
+    transcript_languages = serializers.SlugRelatedField(
+        many=True, read_only=True, slug_field='code',
+        help_text=_('Languages that course runs in this program have available transcripts in.'),
+    )
+    subjects = SubjectSerializer(many=True)
+    staff = PersonSerializer(many=True)
+    start = serializers.DateTimeField()
```
Comment (clintonb) [inline course_discovery/apps/api/serializers.py]: Remove this field. As mentioned in another comment, it does not need to be exposed.

**Comment by clintonb** [comment]
👍 once start field removed from serializer and tests.

File: course_discovery/apps/api/serializers.py (hunk 9 above)
Comment (mikedikan) [reply course_discovery/apps/api/serializers.py]: Just saw your other comment - i'll remove the 'start' field.

**Comment by mikedikan** [comment]
all outstanding comments addressed - waiting for tests to pass then merging. 🎉
